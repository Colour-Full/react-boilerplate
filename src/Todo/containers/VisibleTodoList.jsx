// @flow
import { connect } from 'react-redux'
import { type Dispatch } from 'redux'
import { getTodoListFromState, getVisibilityFilterFromState } from '../api-todo/connectors/todos'
import { toggleTodo, type ToggleTodoAction } from '../api-todo/actions/toggleTodo'
import { type AppState } from 'src/reducers'
import { VisibilityFilters, type VisibilityState } from '../api-todo/reducers/visibilityFilter'
import TodoListComponent from '../components/TodoList'
import { type TodoList, type TodoTypeFromReducer } from '../api-todo/reducers/todo'

export const getVisibleTodos = (todos: TodoList, filter: VisibilityState): TodoList | Error => {
  switch (filter) {
    case VisibilityFilters.SHOW_ALL:
      return todos
    case VisibilityFilters.SHOW_COMPLETED:
      return todos.filter((todo: TodoTypeFromReducer): boolean => todo.completed)
    case VisibilityFilters.SHOW_ACTIVE:
      return todos.filter((todo: TodoTypeFromReducer): boolean => !todo.completed)
    default:
      throw new Error('Unknown filter: ' + filter)
  }
}

type VisibleTodosProp = {
  todos: TodoList | Error
}

export const mapStateToProps = (state: AppState): VisibleTodosProp => {
  // we get the data we want from the state via the connectors
  // which may have been transformed into a specific structure
  // but not processed or filtered etc. That work is done here..
  const todos = getTodoListFromState(state)
  const filter = getVisibilityFilterFromState(state)

  return {
    todos: getVisibleTodos(todos, filter)
  }
}

type ToggleTodoProp = {
  toggleTodo: (id: number) => ToggleTodoAction
}

export const mapDispatchToProps = (dispatch: Dispatch<*>): ToggleTodoProp => ({
  toggleTodo: (id: number): ToggleTodoAction => dispatch(toggleTodo(id))
})

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(TodoListComponent)
