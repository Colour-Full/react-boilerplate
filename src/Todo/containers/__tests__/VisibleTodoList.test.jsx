import { getVisibleTodos, mapStateToProps, mapDispatchToProps } from '../VisibleTodoList'

describe('A VisibleTodoList container', () => {
  it('Should have getVisibleTodos defined', () => {
    expect(getVisibleTodos).toBeDefined()
  })
  it('getVisibleTodos should throw an error if no valid filter is passed', () => {
    const todos = [
      { id: 0, text: 'First Thing To Do', completed: true },
      { id: 1, text: 'Second Thing To Do', completed: false },
      { id: 2, text: 'Third Thing To Do', completed: false }
    ]
    const filter = {
      SHOW_ALL: 'SHOW_ALL',
      SHOW_COMPLETED: 'SHOW_COMPLETED',
      SHOW_ACTIVE: 'SHOW_ACTIVE',
      ERROR: 'ERROR'
    }
    const error = 'Unknown filter: ERROR'
    expect(() => { getVisibleTodos(todos, filter.ERROR) }).toThrow(error)
  })
  it('getVisibleTodos should handle SHOW_ALL filter', () => {
    const todos = [
      { id: 0, text: 'First Thing To Do', completed: true },
      { id: 1, text: 'Second Thing To Do', completed: false },
      { id: 2, text: 'Third Thing To Do', completed: false }
    ]
    const filter = {
      SHOW_ALL: 'SHOW_ALL',
      SHOW_COMPLETED: 'SHOW_COMPLETED',
      SHOW_ACTIVE: 'SHOW_ACTIVE',
      ERROR: 'ERROR'
    }
    expect(getVisibleTodos(todos, filter.SHOW_ALL)).toMatchSnapshot()
  })
  it('getVisibleTodos should handle SHOW_ACTIVE filter', () => {
    const todos = [
      { id: 0, text: 'First Thing To Do', completed: true },
      { id: 1, text: 'Second Thing To Do', completed: false },
      { id: 2, text: 'Third Thing To Do', completed: false }
    ]
    const filter = {
      SHOW_ALL: 'SHOW_ALL',
      SHOW_COMPLETED: 'SHOW_COMPLETED',
      SHOW_ACTIVE: 'SHOW_ACTIVE',
      ERROR: 'ERROR'
    }
    expect(getVisibleTodos(todos, filter.SHOW_ACTIVE)).toMatchSnapshot()
  })
  it('getVisibleTodos should handle SHOW_COMPLETED filter', () => {
    const todos = [
      { id: 0, text: 'First Thing To Do', completed: true },
      { id: 1, text: 'Second Thing To Do', completed: false },
      { id: 2, text: 'Third Thing To Do', completed: false }
    ]
    const filter = {
      SHOW_ALL: 'SHOW_ALL',
      SHOW_COMPLETED: 'SHOW_COMPLETED',
      SHOW_ACTIVE: 'SHOW_ACTIVE',
      ERROR: 'ERROR'
    }
    expect(getVisibleTodos(todos, filter.SHOW_COMPLETED)).toMatchSnapshot()
  })
  it('Should have mapStateToProps defined', () => {
    expect(mapStateToProps).toBeDefined()
  })
  it('mapStateToProps should return all todos prop when called with state with visibility set to `SHOW_ALL`', () => {
    const state = {
      todos: [
        { id: 0, text: 'First Thing To Do', completed: true },
        { id: 1, text: 'Second Thing To Do', completed: false },
        { id: 2, text: 'Third Thing To Do', completed: false }
      ],
      visibility: 'SHOW_ALL'
    }
    expect(mapStateToProps(state)).toMatchSnapshot()
  })
  it('mapStateToProps should return all todos prop when called with state with visibility set to `SHOW_ALL`', () => {
    const state = {
      todos: [
        { id: 0, text: 'First Thing To Do', completed: true },
        { id: 1, text: 'Second Thing To Do', completed: false },
        { id: 2, text: 'Third Thing To Do', completed: false }
      ],
      visibility: 'SHOW_ALL'
    }
    expect(mapStateToProps(state)).toMatchSnapshot()
  })
  it('mapStateToProps should return todos prop containing `all todos` when called with state with visibility set to `SHOW_ACTIVE`', () => {
    const state = {
      todos: [
        { id: 0, text: 'First Thing To Do', completed: true },
        { id: 1, text: 'Second Thing To Do', completed: false },
        { id: 2, text: 'Third Thing To Do', completed: false }
      ],
      visibility: 'SHOW_ALL'
    }
    expect(mapStateToProps(state)).toMatchSnapshot()
  })
  it('mapStateToProps should return todos prop containing only `active todos` when called with state with visibility set to `SHOW_ACTIVE`', () => {
    const state = {
      todos: [
        { id: 0, text: 'First Thing To Do', completed: true },
        { id: 1, text: 'Second Thing To Do', completed: false },
        { id: 2, text: 'Third Thing To Do', completed: false }
      ],
      visibility: 'SHOW_ACTIVE'
    }
    expect(mapStateToProps(state)).toMatchSnapshot()
  })
  it('mapStateToProps should return todos prop containing only `completed todos` when called with state with visibility set to `SHOW_COMPLETED`', () => {
    const state = {
      todos: [
        { id: 0, text: 'First Thing To Do', completed: true },
        { id: 1, text: 'Second Thing To Do', completed: false },
        { id: 2, text: 'Third Thing To Do', completed: false }
      ],
      visibility: 'SHOW_COMPLETED'
    }
    expect(mapStateToProps(state)).toMatchSnapshot()
  })
  it('Should have mapDispatchToProps defined', () => {
    expect(mapDispatchToProps).toBeDefined()
  })
  it('mapDispatchToProps defined', () => {
    expect(mapDispatchToProps).toBeDefined()
  })
  it('mapDispatchToProps should return toggleTodo prop when called with dispatch', () => {
    const dispatch = jest.fn()
    const toggleTodo = jest.fn()
    const result = {
      toggleTodo: (id) => dispatch(toggleTodo(id))
    }
    expect(JSON.stringify(mapDispatchToProps(dispatch))).toEqual(JSON.stringify(result))
  })
})
