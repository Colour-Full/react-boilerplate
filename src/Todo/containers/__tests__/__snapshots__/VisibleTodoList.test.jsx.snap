// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A VisibleTodoList container getVisibleTodos should handle SHOW_ACTIVE filter 1`] = `
Array [
  Object {
    "completed": false,
    "id": 1,
    "text": "Second Thing To Do",
  },
  Object {
    "completed": false,
    "id": 2,
    "text": "Third Thing To Do",
  },
]
`;

exports[`A VisibleTodoList container getVisibleTodos should handle SHOW_ALL filter 1`] = `
Array [
  Object {
    "completed": true,
    "id": 0,
    "text": "First Thing To Do",
  },
  Object {
    "completed": false,
    "id": 1,
    "text": "Second Thing To Do",
  },
  Object {
    "completed": false,
    "id": 2,
    "text": "Third Thing To Do",
  },
]
`;

exports[`A VisibleTodoList container getVisibleTodos should handle SHOW_COMPLETED filter 1`] = `
Array [
  Object {
    "completed": true,
    "id": 0,
    "text": "First Thing To Do",
  },
]
`;

exports[`A VisibleTodoList container mapStateToProps should return all todos prop when called with state with visibility set to \`SHOW_ALL\` 1`] = `
Object {
  "todos": Array [
    Object {
      "completed": true,
      "id": 0,
      "text": "First Thing To Do",
    },
    Object {
      "completed": false,
      "id": 1,
      "text": "Second Thing To Do",
    },
    Object {
      "completed": false,
      "id": 2,
      "text": "Third Thing To Do",
    },
  ],
}
`;

exports[`A VisibleTodoList container mapStateToProps should return all todos prop when called with state with visibility set to \`SHOW_ALL\` 2`] = `
Object {
  "todos": Array [
    Object {
      "completed": true,
      "id": 0,
      "text": "First Thing To Do",
    },
    Object {
      "completed": false,
      "id": 1,
      "text": "Second Thing To Do",
    },
    Object {
      "completed": false,
      "id": 2,
      "text": "Third Thing To Do",
    },
  ],
}
`;

exports[`A VisibleTodoList container mapStateToProps should return todos prop containing \`all todos\` when called with state with visibility set to \`SHOW_ACTIVE\` 1`] = `
Object {
  "todos": Array [
    Object {
      "completed": true,
      "id": 0,
      "text": "First Thing To Do",
    },
    Object {
      "completed": false,
      "id": 1,
      "text": "Second Thing To Do",
    },
    Object {
      "completed": false,
      "id": 2,
      "text": "Third Thing To Do",
    },
  ],
}
`;

exports[`A VisibleTodoList container mapStateToProps should return todos prop containing only \`active todos\` when called with state with visibility set to \`SHOW_ACTIVE\` 1`] = `
Object {
  "todos": Array [
    Object {
      "completed": false,
      "id": 1,
      "text": "Second Thing To Do",
    },
    Object {
      "completed": false,
      "id": 2,
      "text": "Third Thing To Do",
    },
  ],
}
`;

exports[`A VisibleTodoList container mapStateToProps should return todos prop containing only \`completed todos\` when called with state with visibility set to \`SHOW_COMPLETED\` 1`] = `
Object {
  "todos": Array [
    Object {
      "completed": true,
      "id": 0,
      "text": "First Thing To Do",
    },
  ],
}
`;
